torch/csrc/autograd/engine.cpp

```diff
@@ -742,44 +755,115 @@ auto Engine::execute(const edge_list& roots,
   return execute_with_graph_task(graph_task, graph_root)->wait();
 }
 
-void Engine::initialize_threads_pool() {
+void Engine::initialize_device_threads_pool() {
   track_bad_autograd_forks();
   TORCH_CHECK(!in_bad_autograd_fork,
               "Unable to handle autograd's threading in combination with fork-based multiprocessing. "
               "See https://github.com/pytorch/pytorch/wiki/Autograd-and-Fork");
-  std::call_once(start_threads_flag_, &Engine::start_threads, this);
+  std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
 }
 
 void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
-  initialize_threads_pool();
-  ready_queue(at::kCPU).push(
+  initialize_device_threads_pool();
+  std::shared_ptr<GraphTask> graph_task = task.base_.lock();
+  // The graph_task must be alive at this point, because internal autograd machinary
+  // who calls this API (Distributed Autograd Engine) increases outstanding_tasks_
+  // outside this API to keep the GraphTask alive
+  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
+  ready_queue(graph_task, at::kCPU).push(
       std::move(task), /* incrementOutstandingTasks */ false);
 }
 
+void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTask>& graph_task) {
+  std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
```

Comment:
You should restore the thread local state here no? Otherwise the reentrant detection might not work depending on the worker thread that you got from the thread pool.

Reply:
for a similar reason as https://github.com/pytorch/pytorch/pull/33157/files#r393917628, reentrant detection in the async mode will just create its own ReadyQueue and execute using that.

Reply:
I don't think we want to do that. That could lead to big regression in speed for this code:
```
inp = torch.rand(10, requires_grad=True)

# I have two independent branches.
# Run the small one on CPU as it's faster there and the big one one GPU.
model1 = MySmallModel()
model2 = MyBigModel().cuda()

out_cpu = model1(inp)
out_gpu = checkpoint(model2, inp) # Use checkpoint to avoid running out of GPU memory
res = torch.cat([out_cpu, out_gpu.cpu()], 0)
crit(res, gt).backward()
```
In this case, the reentrant due to the checkpoint will completely block the CPU queue until the GPU thread is done. But there is a lot of CPU work available that it could have done in the meantime.

Reply:
Regarding reentrant detection, IIUC we can only have one reentrant call in this codepath right? If we have a reentrant backward, that would end up calling the Engine::execute() and execute the sync codepath right? The first call to the sync codepath wouldn't detect this is a reentrant call, but subsequent reentrant calls in that codepath would detect this right?

@albanD Regarding your example, that wouldn't even trigger this codepath since it would all use the sync codepath?

Reply:
If the main backward using the distributed engine, then it will trigger this code right?

Subsequent reentrant would be ok yes.
But the first one is still not. And this example is about the first one being problematic.

Reply:
@albanD Yes, if we use the distributed engine, this would be the case. My point was that this wouldn't cause regression for any existing code that uses local backward. For distributed engine, it's very unlikely we're doing something like this (we're sure we don't do this for our internal use case). Since this is a perf optimization, can we improve this in a follow up PR? Would prefer to land this PR soon and incrementally improve things (unless there are correctness issues).

Reply:
Ho right I see your point that there won't be any regression for current use case! 👍 

> it's very unlikely we're doing something like this

I know that the `checkpoint` module is widely used at FAIR, but maybe for CPU training it is less useful indeed.

> Since this is a perf optimization

From the pure engine API it is a bit weird that if you run in async mode or not, you get a completely different behavior. But I guess not one should be using that API anyway :D

Reply:
Thanks @albanD @pritamdamania for the discussion! Since this is not regressing any local autograd engine perf use cases, I will do a follow up for the dist engine of that perf optimization, created a issue to track all follow ups https://github.com/pytorch/pytorch/issues/35006

> From the pure engine API it is a bit weird that if you run in async mode or not, you get a completely different behavior. But I guess not one should be using that API anyway :D

Yep I agree with it, the `execute_with_graph_task` api is only supposed to be used by internal autograd machinary like distributed autograd. And our C++ API for autograd does not have this function and don't expose this `async_mode` to the user so it should be fine.